include/zephyr/sys/kref.h

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2023 Michael Zimmermann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_KREF_H_
+#define ZEPHYR_INCLUDE_SYS_KREF_H_
+
+#include <stdbool.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/toolchain.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup kref_apis Reference count APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/**
+ * @brief An atomic reference count.
+ *
+ * The main use case is to share the same data between several users and free
+ * it when the count drops to 0. This can also be used on SMP systems.
+ */
+struct k_ref {
+	atomic_t refs;
+};
+
+/** @brief Initialize k_ref at compile-time. */
+#define K_REF_INIT()                                                                               \
+	{                                                                                          \
+		.refs = ATOMIC_INIT(1),                                                            \
+	}
+
+/** @brief Initialize k_ref at runtime. */
+void k_ref_init(struct k_ref *ref);
+
+/**
+ * @brief Increase the reference count.
```

Comment:
I am not sure that these function headers should be using "reference count" as that is an internal implementation detail (I find myself waffling on this).

We can hide that particular detail with phrasing like "Obtain a reference", or "Release a reference", or "If there are no references"

Question ... do you think it would be worthwhile to have a macro or another function whose purpose would be to  retrieve the reference count?

Reply:
This is probably a very fine line given how basic the concept of a "reference counter" is. I wouldn't know how to implement it differently. That being said, I agree with you in that we should remove that terminology if there is a better one - which you've provided.

Personally, I can't think of any reason why you'd want to know the reference count. But then again, Rusts `Rc` has functions like `weak_count` and `strong_count` so there may be a need for some people.

Reply:
I adjusted the documentation